### ISSUE TITLE
How to deal with task placement strategies/constraints?

### ISSUE THREAD

**Issue by CpuID** [dethmix/self-healing-zookeeper#3]
Hi there,

If this topology is used in a multi-tenanted ECS cluster environment (using the same approach of 1 ZK node per ECS service), how do you deal with placement strategies / placement constraints, to ensure no more than 1 ZK node ends up on the same EC2 instance?

My thoughts were `spread` on Task Groups would be ideal, but not yet supported for ECS services. [Requested upstream](https://github.com/aws/containers-roadmap/issues/1103)

Curious on your feedback :)

Thanks.

**Comment by CpuID** [comment]
Aha, in theory, because "host port" and "container port" are defined, each node/service will reserve the host ports 2181/2888/3888 on a given node, which will prevent a second instance making it to the same node?

**Comment by dethmix** [comment]
Hi,
AWS provides placement constraints and one of them is `distinctInstance `. That is exactly what you need. More details can be found at https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-placement-constraints.html

**Comment by CpuID** [comment]
> AWS provides placement constraints and one of them is `distinctInstance `. That is exactly what you need.

`distinctInstance` only applies if you have N x ECS tasks under a single ECS service, to keep them all on separate instances. It won't achieve the desired result when using multiple ECS services, with 1 ECS task per service. The services won't look at eachother's states there...

**Comment by dethmix** [comment]
@CpuID that is true, completely forgot about instance per service setup. But you still can use placement constraints, but with `memberOf` tag. For this particular setup you can mark 3 instances in your cluster as zookeeper hosts. For example for few instances you can set ECS_INSTANCE_ATTRIBUTES into /etc/ecs/ecs.config like following
```
ECS_INSTANCE_ATTRIBUTES={"instance_destination": "zookeeper${zoo_id}"}
```
where zoo_id is ID of your zoo ECS service.
After that for each ECS zoo service you should set following placement_constraints
```
  placement_constraints {
    type       = "memberOf"
    expression = "attribute:instance_destination==zookeeper${count.index+1}"
  }
```
As a result Zoo1 ECS service will always start on instance where instance_destination is zookeeper1. Moreover other services(non zookeeper) still will be able to use these servers if you don't set placement_constraints for them.

**Comment by CpuID** [comment]
Ah, yea ok that would involve setting something specific on specific EC2 instance nodes. Which tends to treat them more like "pets" than "cattle", not a desired outcome for our usage :)

I've opted for a more dynamic approach, and just relied on the host port binding for now to ensure no 2 nodes can be on the same EC2 instance (due to a conflicting port mapping).

I've also requested more support upstream via https://github.com/aws/containers-roadmap/issues/1103 to do this more elegantly (at a service level, rather than an ECS cluster/host level).
